src/twisted/protocols/amp.py

```diff
@@ -674,11 +675,29 @@ def serialize(self):
                 raise TypeError("Unicode value for key %r not allowed: %r" % (k, v))
             if len(k) > MAX_KEY_LENGTH:
                 raise TooLong(True, True, k, None)
-            if len(v) > MAX_VALUE_LENGTH:
+            if len(v) > MAX_VALUE_LENGTH and not version2:
                 raise TooLong(False, True, v, k)
-            for kv in k, v:
-                w(pack("!H", len(kv)))
-                w(kv)
+
+            w(pack("!H", len(k)))
+            w(k)
+
+            if version2:
+                vio = BytesIO(v)
+                del v
+
+                # If the value is an exact multiple of 65535, the last
+                # chunk containing data will be followed by a zero-length chunk
+                # (i.e. when read() returns EOF).
+                chunk = vio.read(MAX_VALUE_LENGTH)
+                while True:
+                    w(pack("!H", len(chunk)))
```

Comment:
I realize that this is coming from the "spec" on amp-protocol.net, but there were never any discussions I saw about how to implement a V2.  Given that this is neither backwards compatible nor has any wire-level negotiation it seems like a pretty bad feature for a version bump.

Reply:
(A better way to have done this would be to normalize the meaning of repeated keys, but I digress.)

Reply:
Well, it is _mostly_ backwards compatible as long as values are smaller than 65534 bytes. Considering it is a breaking change, it should be signaled via a compatibility version change. I would have called AMP v1.1, but AMP v2 already existed conceptually.

Reply:
If you have a key which is just a bit over 64k bytes, you end up screwing up the alternation of keys and values, which means that the failure mode involves the peer hanging forever; that's a pretty bad failure mode.

There's a way to do this which is *entirely* backwards compatible, though; just send the same key repeatedly, or send a key with `.1`, `.2`, etc appended to it.  This ends up adding 0.02% or so worth of overhead to the data (and if you have really large data, you want what's described on https://twistedmatrix.com/trac/ticket/2529 anyway, not just a way to circumvent the length-prefix limit).

You can also do this as a new parameter-type declaration, which also allows you to interrogate your peer's capabilities by checking for an `UNHANDLED` response if you have something you want to upgrade.

If we wanted to do something dramatically better, [netstrings](https://en.wikipedia.org/wiki/Netstring) are the way to go; arguably AMP should have used these in the first place.